components/org.wso2.carbon.identity.application.authenticator.push/src/main/java/org/wso2/carbon/identity/application/authenticator/push/PushAuthenticator.java

```diff
@@ -0,0 +1,495 @@
+     */
+    public static String getConfiguration(AuthenticationContext context, String configName) {
+
+        String configValue = null;
+        if ((context.getProperty(configName)) != null) {
+            configValue = String.valueOf(context.getProperty(configName));
+        }
+        if (log.isDebugEnabled()) {
+            String debugMessage = String.format("Config value for key %s: %s" + configName, configValue);
+            log.debug(debugMessage);
+        }
+        return configValue;
+    }
+
+    @Override
+    protected void processAuthenticationResponse(HttpServletRequest httpServletRequest, HttpServletResponse
+            httpServletResponse, AuthenticationContext authenticationContext) throws AuthenticationFailedException {
+
+        AuthenticatedUser user = authenticationContext.getSequenceConfig().
+                getStepMap().get(authenticationContext.getCurrentStep() - 1).getAuthenticatedUser();
+
+        AuthenticationContext sessionContext = AuthContextCache
+                .getInstance()
+                .getValueFromCacheByRequestId(new AuthContextcacheKey(httpServletRequest
+                        .getParameter("sessionDataKey"))).getAuthenticationContext();
+
+        AuthDataDTO authDataDTO = (AuthDataDTO) sessionContext.getProperty("authData");
+
+        String jwt = authDataDTO.getAuthToken();
+        String serverChallenge = authDataDTO.getChallenge();
+
+        PushJWTValidator validator = new PushJWTValidator();
+        try {
+            if (validateSignature(jwt, serverChallenge)) {
+                String authStatus = validator.getAuthStatus(jwt);
+                // TODO: Change Successful to Allowed
+                if (authStatus.equals("SUCCESSFUL")) {
+                    authenticationContext.setSubject(user);
+                } else if (authStatus.equals("DENIED")) {
+                    String deniedPage = URLEncoder
+                            .encode("/authenticationendpoint/retry.do?status=Authentication Denied!",
+                                    StandardCharsets.UTF_8.name()) + URLEncoder
+                            .encode("&statusMsg=Authentication was denied from the mobile app",
+                                    StandardCharsets.UTF_8.name());
+                    httpServletResponse.sendRedirect(deniedPage);
+                } else {
+                    throw new AuthenticationFailedException("Authentication failed! Auth status unavailable.", user);
+                }
+            } else {
+                authenticationContext.setProperty(PushAuthenticatorConstants.AUTHENTICATION_STATUS, true);
+                throw new AuthenticationFailedException("Authentication failed! Could not verify signature.", user);
+            }
+
+        } catch (IOException e) {
+            log.error("IO exception while trying to validate signature.");
+            throw new AuthenticationFailedException("IO exception while trying to validate the signature", e);
```

Comment:
Add a full stop to the end. 
Lets try to add a bit more context to the error messages that could help identify/differentiate the request that's causing the error.
"Error occurred while trying to validate the signature for device <decided id>" (some non-confidential identifier that can be extracted from the jwt maybe).

Lets try to improve the context information for other error messages as well